pkg/utils/revision/revision_utils.go

```diff
@@ -0,0 +1,368 @@
+	}
+	cr := &appsv1.ControllerRevision{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels:          labelMap,
+			OwnerReferences: []metav1.OwnerReference{*metav1.NewControllerRef(parent, parentKind)},
+			Namespace:       parent.GetNamespace(),
+		},
+		Data:     data,
+		Revision: revision,
+	}
+	hash := HashControllerRevision(cr)
+	cr.Name = ControllerRevisionName(parent.GetName(), hash)
+	if cr.Labels[leaderworkerset.TemplateRevisionHashKey] == "" {
+		cr.Labels[leaderworkerset.TemplateRevisionHashKey] = hash
+	}
+	return cr
+}
+
+// HashControllerRevision hashes the contents of revision's Data using FNV hashing.
+// The returned hash will be a safe encoded string to avoid bad words.
+func HashControllerRevision(revision *appsv1.ControllerRevision) string {
+	hf := fnv.New32()
+	if len(revision.Data.Raw) > 0 {
+		hf.Write(revision.Data.Raw)
+	}
+	if revision.Data.Object != nil {
+		DeepHashObject(hf, revision.Data.Object)
+	}
+	return rand.SafeEncodeString(fmt.Sprint(hf.Sum32()))
+}
+
+// EqualRevision returns true if lhs and rhs are either both nil, if the templateRevisionHash is the same,
+// or if they are semantically equivalent.
+func EqualRevision(lhs *appsv1.ControllerRevision, rhs *appsv1.ControllerRevision) bool {
+
+	if lhs == nil || rhs == nil {
+		return lhs == rhs
+	}
+
+	if lhs.Labels[leaderworkerset.TemplateRevisionHashKey] == rhs.Labels[leaderworkerset.TemplateRevisionHashKey] {
+		return true
+	}
+
+	return bytes.Equal(lhs.Data.Raw, rhs.Data.Raw) && apiequality.Semantic.DeepEqual(lhs.Data.Object, rhs.Data.Object)
+}
+
+type History struct {
+	client.Client
+	context context.Context
+}
+
+// ListControllerRevisions lists all ControllerRevisions matching selector and owned by parent or no other
+// controller. If the returned error is nil the returned slice of ControllerRevisions is valid. If the
+// returned error is not nil, the returned slice is not valid.
+func (h *History) ListControllerRevisions(parent metav1.Object, selector labels.Selector) ([]*appsv1.ControllerRevision, error) {
+	// List all revisions in the namespace that match the selector
+	revisionList := new(appsv1.ControllerRevisionList)
+	err := h.List(h.context, revisionList, client.InNamespace(parent.GetNamespace()), client.MatchingLabelsSelector{Selector: selector})
+	if err != nil {
+		return nil, err
+	}
+	history := revisionList.Items
+	var owned []*appsv1.ControllerRevision
+	for i := range history {
+		ref := metav1.GetControllerOfNoCopy(&history[i])
+		if ref == nil || ref.UID == parent.GetUID() {
+			owned = append(owned, &history[i])
+		}
+
+	}
+	return owned, err
+}
+
+// CreateControllerRevision attempts to create the revision as owned by parent via a ControllerRef. If the returned
+// error is not nil, creation failed. If the returned error is nil, the returned ControllerRevision has been
+// created.
+func (h *History) CreateControllerRevision(parent metav1.Object, revision *appsv1.ControllerRevision) (*appsv1.ControllerRevision, error) {
```

Comment:
As mentioned below, please move this code inside `CreateLeaderWorkerSetRevision` and rename the function to `CreateRevision`